### PR TITLE
Add support for `prettierIgnore` key in `package.json`

### PR DESCRIPTION
**Pull request by mataha**
## Description

Most of what I've wanted to write about had already been discussed in #3460; to keep it brief:

`package.json` is used more and more to store configuration for various development tools (including their gitignore-like files, if any) in lieu of dotfiles. Prettier provides a way to do that for its options, but not for the list of files to ignore itself.

This PR tries to amend that. Implementation is similar to ESLint - via `prettierIgnore` key.

## Checklist

- [x] I’ve added tests to confirm my change works.
- [x] (If changing the API or CLI) I’ve documented the changes I’ve made (in the `docs/` directory).
- [x] (If the change is user-facing) I’ve added my changes to `changelog_unreleased/*/XXXX.md` file following `changelog_unreleased/TEMPLATE.md`.
- [x] I’ve read the [contributing guidelines](https://github.com/prettier/prettier/blob/main/CONTRIBUTING.md).

**✨[Try the playground for this PR](https://prettier.io/playground-redirect)✨**

### CODE REVIEW

**Comment by mataha** [comment]
Anyone?

**Comment by Trim21** [comment]
any progress 👀?

File: src/common/create-ignorer.js [hunk 1]
```diff
@@ -21,9 +31,20 @@ async function createIgnorer(ignorePath, withNodeModules) {
...
 createIgnorer.sync = function (ignorePath, withNodeModules) {
-  const ignoreContent = !ignorePath
+  let ignoreContent = !ignorePath
     ? null
     : getFileContentOrNull.sync(path.resolve(ignorePath));
+
+  if (!ignoreContent) {
+    try {
+      ignoreContent = JSON.parse(
+        getFileContentOrNull.sync(path.resolve("package.json"))
```
Comment (fisker) [inline src/common/create-ignorer.js:41]: I think we should search `package.json` from the file not `cwd`.

Comment (mataha) [reply src/common/create-ignorer.js]: Is something like [this](https://github.com/prettier/prettier/pull/12672/commits/7256153fb0d9bdfd609de8e87ec3111334c290e1) fine?

Comment (fisker) [reply src/common/create-ignorer.js]: No, not the ignore file, the formatting file.

Comment (fisker) [reply src/common/create-ignorer.js]: Maybe it should be done in `getFileInfo` function.

Comment (mataha) [reply src/common/create-ignorer.js]: Okay, how about [this](https://github.com/prettier/prettier/pull/12672/commits/3b0bd49b4b54ef23db2a85a58c8a9ba3db0508c9) then? I have to admit it's pretty ugly though.

File: src/common/get-key-content-or-null.js [hunk 2]
```diff
@@ -0,0 +1,79 @@
+"use strict";
```
Comment (fisker) [inline src/common/get-key-content-or-null.js:1]: Maybe we use cosmiconfig to search for `prettierIgnore`?

```js
cosmiconfig("prettierIgnore", {searchPlaces: ["package.json"]})
```

File: src/cli/format.js [hunk 3]
```diff
@@ -227,8 +227,13 @@ function format(context, input, opt) {
 }
 
 async function createIgnorerFromContextOrDie(context) {
+  const filepath = context.argv.filepath
```
Comment (fisker) [inline src/cli/format.js:230]: Where this filepath come from? There is no `filepath` flag.

Comment (mataha) [reply src/cli/format.js:230]: There is - as [`--stdin-filepath`](https://github.com/prettier/prettier/blob/fa3b95c2e1a3a3351dc1b0168aa2c26f0cc4e9b3/src/main/core-options.js#L97-L106) - and it's used [here](https://github.com/prettier/prettier/blob/fdccc0f48e5bd3a17ae1e3500c1b5c1763c6a368/src/cli/index.js#L99) as well.

File: src/common/get-key-content-or-null.js (hunk 2 above)
Comment (mataha) [reply src/common/get-key-content-or-null.js:1]: My biggest gripe about this is that cosmiconfig is suited for JSON and YAML configuration files, not lists of file globs with or without negations (which ignore files are).

I will try it, however.

**Comment by CiiDyR** [comment]
I'd love to see this land. It would allow us to have pretterignore settings shared across projects. As we can already share settings, but not ignored files (https://prettier.io/docs/en/configuration.html#sharing-configurations)

Is there anything we could contribute to accelerate this?

**Comment by fullstackjedi** [comment]
I would suggest adding this as an option *within* the configuration file (i.e. `.prettierrc.json`). That way, it can be included in shareable configs.

**Comment by nickmccurdy** [comment]
I don't think sharing ignored files would be a good practice, as ignored files tend to vary more based on a project's tooling usage than its style guide, and we wouldn't want to couple something unrelated to style to a config. For reference, ESLint shareable configs also don't allow sharing ignores. If you do have several projects with the same tooling, you can simply copy the `.prettierignore` (or with this, the `package.json`).

**Comment by fullstackjedi** [comment]
> For reference, ESLint shareable configs also don't allow sharing ignores.

Yes, [they do](https://eslint.org/docs/latest/user-guide/configuring/ignoring-code).

File: src/common/get-key-content-or-null.js (hunk 2 above)
Comment (mataha) [reply src/common/get-key-content-or-null.js]: I've changed the implementation - performance is almost the same compared to the previous method while it's a bit nicer on the eyes (I think).

**Comment by mataha** [comment]
> I would suggest adding this as an option _within_ the configuration file (i.e. `.prettierrc.json`). That way, it can be included in shareable configs.

That's a different feature for a different PR.

File: src/common/create-ignorer.js [hunk 4]
```diff
@@ -2,40 +2,80 @@
...
 /**
+ * @param {string} filePath
  * @param {string?} ignorePath
  * @param {boolean?} withNodeModules
  */
-createIgnorer.sync = function (ignorePath, withNodeModules) {
-  const ignoreContent = !ignorePath
-    ? null
-    : getFileContentOrNull.sync(path.resolve(ignorePath));
+createIgnorer.sync = function (filePath, ignorePath, withNodeModules) {
```
Comment (fisker) [inline src/common/create-ignorer.js:28]: Let's change base to `next` branch, so we don't need support sync version.

Comment (mataha) [reply src/common/create-ignorer.js]: Done.

**Comment by ivodolenc** [comment]
It might not be a bad idea to implement this feature directly in the `prettier.config` file.

So when we want to share the settings we have all the options inside one file. This is exactly what `eslint` have.

For example, I can specify in `package.json` custom path for `eslint.config.js` and `prettier.config.js` and clean up the root as much as possible (no need for additional ignore files and custom ignore options inside package.json):

```json
{
  "name": "package.json",
  "type": "module",

  "prettier": "./.config/prettier.config.cjs",
  "eslintConfig": {
    "extends": "./.config/eslint.config.cjs"
  }
}
```

and in one file I have all settings, including `ignorePatterns` like this:

```js
// .config/eslint.config.js

module.exports = {
  root: true,

  extends: [],

  plugins: [],

  ignorePatterns: [
    '.DS_Store',
    'node_modules'
    // ...
  ]
}
```

This would be a huge win to put `ignore` directly inside `prettier` config:

```js
// .config/prettier.config.js

module.exports = {
  singleQuote: true,
  trailingComma: 'none',
  
  plugins: [],

  ignorePatterns: [
    '.DS_Store',
    'node_modules'
    // ...
  ]
}
```

What do you think?

File: src/cli/format.js (hunk 3 above)
Comment (fisker) [reply src/cli/format.js:239]: `filepath` only exists when format stdin, when formatting files, it doesn't exist.

Comment (fisker) [reply src/cli/format.js:239]: See `formatFiles` function, it calls `createIgnorerFromContextOrDie` on first line. I don't think it will work.

**Comment by kachkaev** [comment]
@mataha would you be interested to continue this PR? @ivodolenc's [suggestion](https://github.com/prettier/prettier/pull/12672#issuecomment-1256947760) makes sense – it' be great to see alignment between ESLint and Prettier!

**Comment by aensley** [comment]
I second @ivodolenc's suggestion: https://github.com/prettier/prettier/pull/12672#issuecomment-1256947760

It would be really great to be able to specify `ignorePatterns` in the prettier config.

**Comment by Standard8** [comment]
Note, as this appears to have stalled out, I've started working on adding the options to the configuration. As a first step, I've added command line arguments to echo ESLint - as they were quite easy to do, and I've just posted that in #14223

**Comment by wallpants** [comment]
Is there anything I can do to help get this over the finish line? In a world where we already have a ton of config files in the root of our projects, I'm struggling to understand why we can't specify patterns to ignore in our prettier config file and we're forced to create a new file just for that.

**Comment by kevinwolfcr** [comment]
> Is there anything I can do to help get this over the finish line? In a world where we already have a ton of config files in the root of our projects, I'm struggling to understand why we can't specify patterns to ignore in our prettier config file and we're forced to create a new file just for that.

I think they don't like listening to their users :)

**Comment by Standard8** [comment]
> Is there anything I can do to help get this over the finish line? In a world where we already have a ton of config files in the root of our projects, I'm struggling to understand why we can't specify patterns to ignore in our prettier config file and we're forced to create a new file just for that.

As discussed elsewhere, I think it would be much better to add an option in the `.prettierrc.js` file rather than `package.json`, this would also align it better with ESLint (which I believe was one of the aims).

I tried [doing this previously](https://github.com/prettier/prettier/pull/14223#issuecomment-1401827265) but hit some issues, and despite asking never got any feedback from the developers, even when pinging the devs directly, so I gave up.

Feel free to take my work and try moving it forward though, maybe you'll have more success.

**Comment by mataha** [comment]
I'm in a position in which I can commit some time to this again. I'd like to clarify some things beforehand though.

> @mataha would you be interested to continue this PR? @ivodolenc's [suggestion to use `ignorePatterns`](https://github.com/prettier/prettier/pull/12672#issuecomment-1256947760) makes sense – it' be great to see more alignment between ESLint and Prettier!
> 
> 📖 [Docs on `ingorePatterns` in ESLint config](https://eslint.org/docs/latest/user-guide/configuring/ignoring-code#ignorepatterns-in-config-files)

Again, I believe that's a different feature for a different PR. The core tenet of this is that one can share `prettier` configuration between different projects or even monorepo packages, yet be able to selectively choose which files should be formatted at the project level, e.g. plumbing for package managers, test specifications, generated files... As someone here said, ignorable files tend to be influenced by a project's tooling more than anything - two entities sharing such config might not necessarily use the same package manager, among other things. `ignorePatterns` does not achieve this - it's applied on a configuration level, thus applicable to every package using said config.

That said, the main pain point of this PR continues to be how Prettier should behave upon encountering a `package.json`:

  - stop scanning
  - continue scanning until filesystem root, merge all sections
  - continue scanning if section is malformed or missing

I'll have to rewrite this as several things have changed since then (e.g. ignore facility has moved to `src/utils`).

**Comment by mataha** [comment]
Can't run tests at all due to jestjs/jest#7914.

**Pull request closed without merging**